lib/src/main/java/com/knuddels/jtokkit/GptBytePairEncoding.java

```diff
@@ -36,25 +36,42 @@ final class GptBytePairEncoding implements Encoding {
 
 	@Override
 	public List<Integer> encode(final String text) {
+		return encode(text, null);
+	}
+
+	@Override
+	public List<Integer> encode(final String text, final Integer maxTokens) {
 		for (final String specialToken : specialTokensEncoder.getDecodedTokens()) {
 			if (text.contains(specialToken)) {
 				throw new UnsupportedOperationException("Encoding special tokens is not supported yet.");
 			}
 		}
 
-		return encodeOrdinary(text);
+		return encodeOrdinary(text, maxTokens);
 	}
 
 	@Override
 	public List<Integer> encodeOrdinary(final String text) {
+		return encodeOrdinary(text, null);
+	}
+
+	@Override
+	public List<Integer> encodeOrdinary(final String text, final Integer maxTokens) {
 		final List<Integer> out = new ArrayList<>();
 		final Matcher matcher = pattern.matcher(text);
-		while (matcher.find()) {
+		int tokenCount = 0;
+		while (matcher.find() && maxTokenCountNotReached(maxTokens, tokenCount)) {
 			final ImmutableByteArray match = ImmutableByteArray.from(matcher.group());
 			if (encoder.containsDecodedToken(match)) {
 				out.add(encoder.encode(match));
+				++tokenCount;
 			} else {
-				out.addAll(bytePairMerge(match));
+				List<Integer> tokensToAdd = bytePairMerge(match);
+				if(maxTokenCountExceeded(maxTokens, tokenCount + tokensToAdd.size())) {
```

Comment:
If I am not mistaken, this could lead to the maxTokens not being fully utilized. For example:

Let's say we are at 97 tokens and maxtokens is 100. The next match is `supercalifragilistic` which encodes to 6 tokens, then we would break here and return only 97 tokens

I would expect, that the first 3 tokens (`supercalif`) would also be returned, since I don't believe this library should make any assumptions about semantically sensible splitting

Reply:
@tox-p Your assumption was correct :). I've changed it so that all tokens are returned.